compiler/optimizer/VirtualGuardHeadMerger.cpp

```diff
@@ -185,10 +222,40 @@ static bool safeToMoveGuard(TR::Block *destination, TR::TreeTop *guardCandidate,
              tt->getNode()->getInlinedSiteIndex() != TR::comp()->getInlinedCallSite(guardCandidate->getNode()->getInlinedSiteIndex())._byteCodeInfo.getCallerIndex())))
             return false;
 
+         // Collect privarg store symrefs
+         if (tt->getNode()->chkIsPrivatizedInlinerArg())
+            privArgSymRefs.set(tt->getNode()->getSymbolReference()->getReferenceNumber());
+
          if (tt->getNode()->isNopableInlineGuard()
              && tt->getNode()->getBranchDestination() != branchDest)
             return false;
          }
+
+      // Search for loads of the collected privargs in the taken side of guard1
+      // This is possible if a privarg temp was recycled for the inner call site, with a prior use as an
+      // argument for the outer call site. As the privargs for the inner call site must be evaluated before
+      // both guards, this would result in the recycled temp holding the incorrect value if the guard is ever
+      // taken. Give up if this situation is detected.
+      TR_Stack<TR::Block*> blocksToCheck(TR::comp()->trMemory(), 8, false, stackAlloc);
+      blocksToCheck.push(branchDest->getNode()->getBlock());
+      TR::NodeChecklist checklist(TR::comp());
+      while (!blocksToCheck.isEmpty())
+         {
+         TR::Block *block = blocksToCheck.pop();
+
+         // Search this block for any uses of the privargs we plan to move
+         for (TR::TreeTop *tt = block->getFirstRealTreeTop(); tt->getNode()->getOpCodeValue() != TR::BBEnd; tt = tt->getNextTreeTop())
+            if (searchNodeForUses(tt->getNode(), privArgSymRefs, checklist))
+               return false;
+
+         // Search for any successors that have not merged with the mainline
+         for (auto e = block->getSuccessors().begin(); e != block->getSuccessors().end(); ++e)
```

Comment:
I would prefer to have an itr and an end to match other code and to stash the end to save compile time. Something like:
for (auto itr = block->getSuccessors().begin(), end = block->getSuccessors().end(); itr != end; ++itr)